### PR TITLE
Don't stop containers if there are open sessions

### PR DESCRIPTION
**Pull request by joshbetz**

### CODE REVIEW

**Comment by joshbetz** [comment]
Eventually we may want to watch for the command getting killed (by a closed ssh session, for example) and send the command to update the container count on shutdown so we don't lose track of close containers.

File: src/lib/config.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+const fs = require( 'fs' );
+
+const dir = require( 'os' ).homedir() + '/.vip-cli';
+
+export function get( file, callback ) {
+	try {
+		var data = fs.readFileSync( dir + '/' + file, 'utf8' );
```
Comment (joshbetz) [inline src/lib/config.js]: I tried using `fs.readFile`, but the callback was not firing in some cases.

File: src/lib/sandbox.js [hunk 2]
```diff
@@ -55,13 +56,34 @@ export function runCommand( container, command ) {
...
-	api
-		.post( '/containers/' + container.container_id + '/stop' )
-		.end();
+		if ( ! list[ container.container_name ] ) {
+			list[ container.container_name ] = 1;
+		} else {
+			list[ container.container_name ]++;
+		}
+
+		config.set( 'sbox', list, err => {
```
Comment (nickdaugherty) [inline src/lib/sandbox.js]: We should handle potential errors here.

File: src/lib/sandbox.js [hunk 3]
```diff
@@ -55,13 +56,34 @@ export function runCommand( container, command ) {
...
+			spawn( 'docker', run, { stdio: 'inherit' } );
+
+			config.get( 'sbox', ( err, list ) => {
```
Comment (nickdaugherty) [inline src/lib/sandbox.js]: We should handle potential errors here.

File: src/lib/config.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	} catch (e) {
+		return callback( e );
+	}
+
+	data = JSON.parse( data );
```
Comment (nickdaugherty) [inline src/lib/config.js]: The JSON parsing should be in a `try/catch` then return an error if there is a problem.

File: src/lib/config.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+export function set( file, update = {}, callback ) {
+	callback = callback || function() {};
+
+	fs.readFile( dir + '/' + file, ( err, data ) => {
+		if ( err ) {
+			data = {}
+		} else {
+			data = JSON.parse( data )
```
Comment (nickdaugherty) [inline src/lib/config.js]: The JSON parsing should be in a `try/catch` then return an error if there is a problem.

File: src/lib/sandbox.js [hunk 6]
```diff
@@ -68,11 +72,19 @@ export function runCommand( container, command ) {
...
 			spawn( 'docker', run, { stdio: 'inherit' } );
 
 			config.get( 'sbox', ( err, list ) => {
+				if ( err ) {
+					return console.error( err );
+				}
+
 				list[ container.container_name ]--;
-				config.set( 'sbox', list );
+				config.set( 'sbox', list, err => console.error( err ) );
```
Comment (nickdaugherty) [inline src/lib/sandbox.js]: This will always run without a conditional - what happens when you run `console.error( null )`?

Comment (nickdaugherty) [inline src/lib/sandbox.js]: Also, we probably want to abort here with a `return` if there was an error, right?

Comment (joshbetz) [inline src/lib/sandbox.js]: Oh, good point. I can fix that.

**Comment by nickdaugherty** [comment]
Looks good!

**Comment by david-binda** [comment]
Awesome. Seems working fine. Thank you @joshbetz !

**Pull request merged**